java/ql/src/experimental/Security/CWE/CWE-089/MyBatisAnnotationSqlInjectionLib.qll

```diff
@@ -0,0 +1,128 @@
+/**
+ * Provides classes for SQL injection detection in MyBatis annotation.
+ */
+
+import java
+import MyBatisCommonLib
+import semmle.code.xml.MyBatisMapperXML
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.frameworks.Properties
+
+/** A sink for MyBatis annotation method call an argument. */
+class MyBatisAnnotationMethodCallAnArgument extends DataFlow::Node {
+  MyBatisAnnotationMethodCallAnArgument() {
+    exists(MyBatisSqlOperationAnnotationMethod msoam, MethodAccess ma | ma.getMethod() = msoam |
+      ma.getAnArgument() = this.asExpr()
+    )
+  }
+}
+
+/** Get the #{...} or ${...} parameters in the Mybatis annotation value. */
+private string getAnMybatiAnnotationSetValue(IbatisSqlOperationAnnotation isoa) {
+  result = isoa.getSqlValue().trim().regexpFind("(#|\\$)\\{[^\\}]*\\}", _, _)
+}
+
+predicate isMybatisAnnotationSqlInjection(DataFlow::Node node, IbatisSqlOperationAnnotation isoa) {
+  // MyBatis uses an annotation method to perform SQL operations. This method has only one parameter and
+  // the parameter is not annotated with `@Param`. Improper use of this parameter has a SQL injection vulnerability.
+  // e.g.
```

Comment:
Why are there these 5 cases? They all look very similar: it seems to be

1. Any `${blah}` where we have one parameter which isn't `@Param` annotated
2. Any `${blah}` where any parameter is a Map, List or array
3. Any `${fieldName}`
4. Any `${paramName}` matching an `@Param("paramName")`
5. Any `${parami}` or `${argi}` where the `i`th parameter doesn't have a `@Param` annotation

Cases 3, 4 and 5 make sense. What is special about case 1? Why isn't that taken care of by cases 4 and 5? What about case 2, why do maps / lists / arrays mean we don't need to look for a corresponding `${}`? It looks like maybe there is special `${value}` syntax for a single parameter, but if so why aren't we looking for that syntax in particular? For Lists/Maps/arrays, is there `${param[key]}` or similar syntax? If so shouldn't we be looking for `${paramName[optionalSuffix]}` rather than just `${anything}`?

In general, when IS `${}` syntax safe except for the `getAnMybatisConfigurationVariableKey()` case that is excluded everywhere?

To sum up, we should (a) document this well enough to make it clear why each case exists, and (b) remove any duplication between the cases.

Reply:
Case 1, for example:
 ```java
 @Select(select id,name from test where name like'%${value}%')
Test test(String name);
```
The name variable can be named arbitrarily in the SQL statement. If the number of method variables is greater than 1, it cannot be named arbitrarily.


`Lists/Maps/Arrays` is also a special case. I am still thinking about this part of modeling. Because it is different from `@Param`. `@Param` has a very definite syntax.

Case 1, modeling of a parameter value.
Case 2, Modeling of `Map, List or array`.
Case 3, modeling of entity class fields.
Case 4, modeling of @Param annotated parameter names.
Case 5, modeling of default parameter names.